<br>
<font size=6>Intro to Machine Learning</font>
<br>
<br>
<i>"Надо много учиться, чтобы знать хоть немного".</i><br>
<br>
Первым и обязательным шагом к освоению machine learning должна стать статья <a href='https://vas3k.ru/blog/machine_learning/'>Машинное обучение для людей</a>.<br>
Именно с неё я советую каждому начать изучение. Она даст необходимое понимание и представление о направление ML.<br> 
Статья большая и не обязательно читать её за один раз, посмотрите на несколько алгоритмов и можете возвращаться сюда.<br>
Но после следует обязательно дочитать статью.<br><br>
  

Материалы занятия:
- Курс по machine learning на kaggle https://www.kaggle.com/learn

<br>
<font size=6>Эпизод Ⅰ</font><br>
<br>
<font size=5>First model</font>
<hr>
Возьмем датасет <a href='https://www.kaggle.com/datasets/dansbecker/melbourne-housing-snapshot'>Melbourne Housing Snapshot</a> и будем предсказывать цены на жилье.<br>
<br>

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('data/melb_data.csv')
data.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

<br>
Мы говорили, что при работе с данными важно понимать с какими признаками мы работаем.<br>
Иногда в них бывает нехватка и начинаются проблемы, но зачастую при определение нашей задачи, некоторые признаки бывают лишними либо влияют на нашу задачу сильнее других. Например, <b>CouncilArea: Управляющий совет области</b> не так влияет на задачу предсказывать цену жилья, нежели признак <b>Rooms: количество комнат</b>.<br>
Для определения наиболее важных признаков применяют различные статистические методы, о них мы узнаем позже, сейчас же попробуем интуиктивно отобрать полезные признаки.<br>
<br>

In [3]:
data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

<br>
Нам следует подробней изучить каждый признак.<br>
Но для простаты мы отлижим разведывательный анализ и просто посмотрим на пропуски и описание.<br>
<br>

In [4]:
data.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                62
Landsize            0
BuildingArea     6450
YearBuilt        5375
CouncilArea      1369
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

<br>
В данных Мельбурна есть некоторые отсутствующие значения.<br>
Возьмем самый простой вариант для работы с пропусками и выкинем дома из наших данных:<br>
<br>

In [5]:
data = data.dropna(axis=0)

In [6]:
data.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
CouncilArea      0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

<br>
Хорошо, данные обработаны и условно проанализированы<br>
<br>
Наша задача в предсказании цены или признака <b>price</b>.<br>
Такой признак называют целевой переменной(признак) и обычно записывают в перменную y<br>
<br>

In [7]:
y = data.Price

<br>
Большинство алгоритмов принимают только числовые признаки.<br>
Возьмем наиболее важные числовые признаки<br>
<br>

In [8]:
data_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']
X = data[data_features]

Датасет, по которому будет обучаться наш алгоритм, записывается в переменную X

In [9]:
X.describe()

Rooms     Bathroom      Landsize    Lattitude   Longtitude
count  6196.000000  6196.000000   6196.000000  6196.000000  6196.000000
mean      2.931407     1.576340    471.006940   -37.807904   144.990201
std       0.971079     0.711362    897.449881     0.075850     0.099165
min       1.000000     1.000000      0.000000   -38.164920   144.542370
25%       2.000000     1.000000    152.000000   -37.855438   144.926198
50%       3.000000     1.000000    373.000000   -37.802250   144.995800
75%       4.000000     2.000000    628.000000   -37.758200   145.052700
max       8.000000     8.000000  37000.000000   -37.457090   145.526350

In [10]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

<br>
<font size=4><b>Building Model</b></font><br>
<br>
Мы будем использовать библиотеку <b>scikit-learn</b> для создания своих моделей. При кодировании эта библиотека записывается как sklearn, как вы увидите в примере кода. Scikit-learn — это, пожалуй, самая популярная библиотека в ML.<br>
<br>
Шаги построения и использования модели:

 - Define: какой тип модели это будет? Дерево решений? Какой-то другой тип модели?
 - fit: Тренировка или обучение модели. Это сердце ML.
 - Predict: получить прогноз от модели
 - Evaluate: оценить насколько точны прогнозы модели.
    
    
 <br>
 Вот пример определения модели "дерево решений" с помощью scikit-learn и ее сопоставления с функциями и целевой переменной:

In [11]:
!pip install scikit-learn

In [12]:
from sklearn.tree import DecisionTreeRegressor

# Определение модели. Укажите число для random_state, чтобы обеспечить одинаковые результаты при каждом запуске.
model = DecisionTreeRegressor(random_state=1)
# Обучение модели
model.fit(X, y)

DecisionTreeRegressor(random_state=1)

<br>
Многие модели машинного обучения допускают некоторую случайность при обучении модели. Указание числа для random_state гарантирует получение одинаковых результатов при каждом запуске. Это считается хорошей практикой. Вы используете любое число, и качество модели не будет существенно зависеть от того, какое именно значение вы выберете.<br>
<br>
Теперь у нас есть подогнанная модель, которую мы можем использовать для прогнозирования.<br>
<br>
На практике вы захотите делать прогнозы для новых домов, которые появятся на рынке, а не для домов, на которые у нас уже есть цены. Но мы сделаем прогнозы для первых нескольких строк обучающих данных, чтобы увидеть, как работает функция прогнозирования.<br>
<br>

In [13]:
print("Предсказание цены для домов:")
print(X.head())
print("Цена:")
predict = list(model.predict(X.head()))
print(predict)

Предсказание цены для домов:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
Цена:
[1035000.0, 1465000.0, 1600000.0, 1876000.0, 1636000.0]


<br>
<font size=6>Эпизод Ⅱ</font><br>
<br>
<font size=5>Model Validation</font>
<hr>
Модель построена и готова предсказывать цены, но как узнать хорошо ли она их предсказывает?<br>
<br>
В этом эпизоде мы научимся измерять качество модели. Такой показатель является основным для улучшения работы модели.<br>
Первая мысль, которая могла вас посетить, так это взять набор с домами и их ценами, закинуть дома в обученную модель, но не говорить ей цены.<br>
Она предскажет цену для каждого дома и мы просто сравним с фактической ценой. Примерно так все и работает. Но с этим есть множество проблем, <br>
одна из них - для 10 000 домов вы не захотите проверять схожесть цены и предсказания, плюс к этому забудете сколько было хороший, а сколько плохих.<br>
Такая проблема решается обобщенной оценкой.<br>
<br>
Существует много показателей для обобщения качества модели, но мы начнем с показателя, называемого <b>средней абсолютной ошибкой (также называемого MAE)</b>.<br>
Давайте разберем эту метрику, начиная с последнего слова «ошибка».<br>
<br>
Ошибка прогноза для каждого дома:<br>
<br>
ошибка = фактический прогноз<br>
Итак, если дом стоит 150 000 долларов, а вы предсказали, что он будет стоить 100 000 долларов, ошибка составит 50 000 долларов.<br>
<br>
С метрикой MAE мы берем абсолютное значение каждой ошибки. Это преобразует каждую ошибку в положительное число. Затем мы берем среднее значение этих абсолютных ошибок.<br>
<br>
В среднем наши прогнозы отличаются примерно на Х.<br>
<Br>

In [14]:
X

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1          2       1.0     156.0  -37.80790   144.99340
2          3       2.0     134.0  -37.80930   144.99440
4          4       1.0     120.0  -37.80720   144.99410
6          3       2.0     245.0  -37.80240   144.99930
7          2       1.0     256.0  -37.80600   144.99540
...      ...       ...       ...        ...         ...
12205      3       2.0     972.0  -37.51232   145.13282
12206      3       1.0     179.0  -37.86558   144.90474
12207      1       1.0       0.0  -37.85588   144.89936
12209      2       1.0       0.0  -37.85581   144.99025
12212      6       3.0    1087.0  -37.81038   144.89389

[6196 rows x 5 columns]

In [15]:
from sklearn.metrics import mean_absolute_error


predict = model.predict(X)
mean_absolute_error(y, predict)

1115.7467183128902

<br>
<b>Средняя абсолютная ошибка</b> составляет 1115 долларов, довольно неплохой результат для стоимости дома.<br>
Но есть проблема: <i>при обучение и проверки использовался один и тот же набор данных.</i><br>
<br>
Тоже самое, что боксер будет тренироваться каждый день с боксерами из своей команды и в день соревнований выйдет на бой с одним из них.<br>
Его шансы победить(показать высокий результат) необычайно высоки. Но дать ему соперника из другой команды, шансы победить уже не такие высокие.<br>
<br>
Точно также работают данные:<br> 
Представьте себе, что на большом рынке недвижимости цвет двери не связан со стоимостью дома.<br>
<br>
Однако в выборке данных, которую вы использовали для обучения модели, все дома с зелеными дверями были очень дорогими. Задача модели состоит в том, чтобы в наборе данных найти закономерности, которые предсказывают цены на жилье, поэтому она увидит эту закономерность и всегда будет предсказывать высокие цены на дома с зелеными дверями.<br>
<br>
Поскольку этот шаблон(зеленные двери = дорогой дом) был получен из обучающих данных, модель будет выглядеть точной в обучающих данных.<br>
<br>
Но если этот шаблон не выполняется, когда модель видит новые данные, модель будет очень неточной при использовании на практике.<br>
<br>
<br>
Поскольку ценность модели заключается в прогнозировании новых данных, важно проверять её работу на данных, которые не участвовали в обучение.<br>
Данные - самая важная часть ML, они приносят нам немало проблем, одной из которых является их нехватка. Нам нужны данные, которые не будут участвовать в обучение.
Ехать в Мельбурн и собирать новые - не вариант. Самым простым способом будем разбить имеющиеся данные на <b>тренировачные</b> и <b>тестирующие</b><br>
<br>
В библиотеке scikit-learn есть функция train_test_split для разбиения данных на две части. Мы будем использовать некоторые из этих данных в качестве обучающих данных для соответствия модели, а другие данные будем использовать в качестве проверочных данных для вычисления mean_absolute_error.<br>
<br>

In [16]:
from sklearn.model_selection import train_test_split


train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)
model = DecisionTreeRegressor()
model.fit(train_X, train_y)
predict = model.predict(val_X)
print(mean_absolute_error(val_y, predict))

255540.16785022596


In [17]:
data.Price.mean() / mean_absolute_error(val_y, predict) # отношение средней цены на ошибку

4.18262307275425

<br>
Ошибка сильно изменилась. В этом разница между почти совершенно правильной моделью и моделью, непригодной для большинства практических целей. Для справки: средняя стоимость дома в данных проверки составляет 1,1 миллиона долларов. Таким образом, ошибка в новых данных составляет около четверти средней стоимости дома.<br>
<br>

<br>
<font size=6>Эпизод III</font><br>
<br>
<font size=5>Underfitting and Overfitting</font>
<hr>
Модель обучается, используя набор данных, который мы ей дали. <br>
Что означает <i>обучается на данных?</i><br>
Модель пытается найти закономернности в данных, понять почему тот или иной дом стоит столько сколько стоит.<br>
Она смотрим на признаки и сопоставляет их с целевой переменной - ценой. <br>
Важно при этом не дать ей <b>переобучиться(overfitting)</b>, если позволить ей "долго" изучать данные, она найдет закономерности, которые ничего общего с нашей задачей не имеют.<br>
При переобучение модель отлично показывает себя на обучающих данных, но при это плохо на данных проверки. Почему так?<br>
Дело в том, что теперь модель не умеет обобщать свои данные, ей нужно узнать про исследуемый объект ту информацию, которая была в обучение. Например, при обучение она всегда смотрела на улицу, где находится дом, но теперь на реальном наборе данных, где нет улицы или все улицы одинаковые, она выдает плохие результаты, потому что не понимает как оценить стоимость дома, если нет улиц, которые она видела раньше. Когда мы читаем информацию про дом, мы примерно можем обобщить его цену, посмотрев на кол-во комнат, площадь, район, но что если бы мы всю жизнь видели дома только одного вида? У нас не было бы возможности сравнивать и оценивать новые дома. Хороший пример с животные есть <a href='https://newtechaudit.ru/overfitting/'>здесь</a>.<br>
<br>
Что делать? Ну, приходится либо увеличивать разнообразие данных, либо ограничивать обучение. Увеличение данных - это всегда хорошо, но не всегда возможно. Тогда мы ограничим обучение модели и будем получать обобщенные выводы. Но тут возникает проблема со слишком обобщенными выводами. Модель настолько плохо выучила свой датасет, что ничего не знает и выдает результаты с большой ошибкой - это называется <b>недооснащение</b> или <b>недообучение</b>, что уже не так круто звучит, лучше уже просто <b>underfitting</b>.<br>
<br>
Необходимо искать середину между <b>underfitting и overfitting</b>. Поиск нужной точки происходит на графике:<br>
<br>
<img src='data/u-o.png' width="700" height="500">
<br>
<br>
На графике зависимость MAE от глубины алгоритма "дерево решений". Глубину можно задавать в качестве аргумента.<br>
Рассмотрим синию линию - <b>training</b>. Линия показывает величину ошибки на обучающем наборе данных, видно, что ошибка с ростом глубины падает и все с ней хорошо.<br>
Но если посмотреть на красную линию - <b>validation</b>, которая показывает ошибку на тестовом наборе, она ведет себя как парабола, с самого начала падает, а потом возрастает.<br>
Вот как раз в точне минимума, мы должны остановиться и взять ту глубину алгоритма, которая нам нужна.<br>
<br>

In [54]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor


def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes)
    model.fit(train_X, train_y)
    predict = model.predict(val_X)
    mae = mean_absolute_error(predict, val_y)
    print(f'leaf nodes: {max_leaf_nodes}   MAE: {mae}')

for leaf in [5, 50, 500, 5000, 50000]:
    get_mae(leaf, train_X, val_X, train_y, val_y)

leaf nodes: 5   MAE: 369673.0400167675
leaf nodes: 50   MAE: 266644.21831092256
leaf nodes: 500   MAE: 239476.85556340555
leaf nodes: 5000   MAE: 253125.2872821175
leaf nodes: 50000   MAE: 253261.35958683022


<br>
Лучший результат показывает количество листьев = 500, после уже идет переобучение.<br>
<br>

<br>
<font size=6>Эпизод IV</font><br>
<br>
<font size=5>Random Forest</font>
<hr>
Случайный лес использует множество деревьев и делает прогноз, усредняя прогнозы каждого дерева компонентов. Как правило, он имеет гораздо лучшую точность прогнозирования, чем одно дерево решений, и хорошо работает с параметрами по умолчанию. Если вы продолжите моделирование, вы сможете изучить больше моделей с еще более высокой производительностью, но многие из них чувствительны к получению правильных параметров.<br>
<br>

In [62]:
from sklearn.ensemble import RandomForestRegressor


model = RandomForestRegressor(random_state=0)
model.fit(train_X, train_y)
predict = model.predict(val_X)
mae = mean_absolute_error(predict, val_y)
print(mae)

192287.89755541208


<br>
Есть параметры, которые позволяют изменить производительность случайного леса так же, как мы изменили максимальную глубину одиночного дерева решений. Но одна из лучших особенностей моделей Random Forest заключается в том, что они обычно работают разумно даже без этой настройки.<br>
<br>
SimpleImputer для замены отсутствующих значений средним значением по каждому столбцу.

Хотя это просто, заполнение среднего значения обычно работает достаточно хорошо (но это зависит от набора данных). Хотя статистики экспериментировали с более сложными способами определения вмененных значений (например, вменение регрессии), сложные стратегии обычно не дают дополнительных преимуществ после включения результатов в сложные модели машинного обучения.